Suppress Generic error on inocuous stage change

dhilst opened the pull request:
Is expected that stages don't change in decreasing direction so suppress the
error message entirely.

<!--
Any PR related to Mautic 2 issues is not relavant anymore, please consider upgrading your code to the Mautic 3 series (staging/3.0 branch).
-->
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | staging for features or enhancements / 3.2 for bug fixes <!-- see below -->
| Bug fix?                               | yes
| New feature?                           |no
| Deprecations?                          |None
| BC breaks?                             | No
| Automated tests included?              | No
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#step-5-work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the staging branch.
-->

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do.
-->
#### Description:

Stage change events fail with "Generic error" when the contact stage is > then the new stage. This is the expected behavior but the "Generic error" message is misleading. I faced this problem and talk a little about in the [slack](https://mautic.slack.com/archives/C02GEN0SG/p1605816149414600)

<!--
If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to reproduce the error:
To reproduce the generic error just

1. Create two stages, A, B with B weight > than A weight
2. Create a contact and set it's stage to B
3. Create a campaign, add a sole change stage action and set stage to A
4. Add the contact to the campaign and run `mautic:campaign:trigger` command
5. Checkout the contact events log, there is a Generic error 
<img width="781" alt="image" src="https://user-images.githubusercontent.com/392820/99738029-5abadc80-2aa8-11eb-9638-207525d4d404.png">


#### Steps to test this PR:
1. Load up this PR
2. Reproduce the exactly the same steps above

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->


Side note on tests:

I didn't included any automated test, how/where can I do it?

Regards,

kuzmany commented:
Could we add also reason to failed? 
Something like contact cannot move to stage with lower weight.

dhilst commented:
Yes, I did a test with this regarding, but setFailed set an error message and I find it confusing, my reasoning was something like:

- If this is an error how do I fix it?
- You don't fix is the expected behavior
- So is not an error

Anyway, I will update the PR so we can work on examples

kuzmany commented:
@dhilst if it setup just `setFailed()` what is resuld in timeline? Just log without error? Than could be better and I let it as you push

dhilst commented:
@kuzmany  here is a screenshot of the timeline with `setFailed()`

<img width="785" alt="image" src="https://user-images.githubusercontent.com/392820/99807734-b031d080-2b1e-11eb-9425-a2a502dfe98d.png">


here is a screenshot of the timeline with `setFailed(<translated message>)`

<img width="800" alt="image" src="https://user-images.githubusercontent.com/392820/99807856-d5264380-2b1e-11eb-9ca4-fca74430f4f8.png">

(plz ignore the _then_, it was supposed to be _than_)

dhilst commented:
In 3.1.2 I'm getting this messages 
<img width="783" alt="image" src="https://user-images.githubusercontent.com/392820/99808183-3f3ee880-2b1f-11eb-8f44-e4707691e567.png">

My fear wit `setFailed()` is that it will not be rescheduled, I'm not sure what would be the right semantic here. If an user is restarted in a campaign will `.setFailed()` prevent the action from being invoked?

kuzmany reviewed (dismissed):
Works 👍

kuzmany commented:
@dhilst Thank you for explain. Then works for me and I approved it.

npracht commented:
@dhilst can you please rebase against staging and also sign the contributor agreement please ?

dhilst commented:
@npracht Sure, how can I sign the contributor agreement? I'm very new to this signing stuff

Forget it I just findout, https://www.mautic.org/contributor-agreement

dhilst committed:
Prevent "Generic error" on Stage change action on UI

Is expected that stages don't change in decreasing direction so suppress the
error message entirely.

dhilst commented:
I had wrong email on my git client, I amended the commit with the right email :)

RCheesley commented:
Hi folks,

As a user, if there has been an error or a reason why a campaign action has unexpectedly not taken effect, I would expect to see the error message telling me why. 

I agree that the Generic Error message is poor, but hiding it isn't the answer in my opinion.

I prefer that we show a message like this:

<img width="800" alt="99807856-d5264380-2b1e-11eb-9ca4-fca74430f4f8" src="https://user-images.githubusercontent.com/2930593/99876819-52ba8400-2bf1-11eb-9d44-02fa19e8a6e6.png">

Maybe it should say 'Unable to change contact stage because current stage B has a greater weight than A' (could we even link to documentation on this? I note that we don't even mention the fact that you're not able to move to a lesser weighted stage in the [documentation](https://docs.mautic.org/en/stages))

kuzmany commented:
Yes, but this is what we discussed above. Not changing stage it's feature. We expect not changing stage, that's not failed action triggering. For me it's ok

RCheesley commented:
I get what you're saying @kuzmany - I can see both sides of this. I've posted the question in the forum to get thoughts from users of Mautic as to whether they would expect to see a warning or not.

RCheesley commented:
@cla-bot check

kuzmany commented:
But I found another PR from Acquia what also change onCampaignTriggerActionChangeStage with more logic:

https://github.com/mautic/mautic/pull/8969

Could you try and let us If works for you?

dhilst commented:
@kuzmany It works for me 👍

kuzmany commented:
Close.
Duplicate  https://github.com/mautic/mautic/pull/8969

kuzmany commented:
@dhilst Can you please approve https://github.com/mautic/mautic/pull/8969
